src/dmd/json.d

```diff
@@ -857,17 +858,165 @@ public:
-    {
-        Module m = (*modules)[i];
-        if (global.params.verbose)
-            message("json gen %s", m.toChars());
-        m.accept(json);
-    }
-    json.generateSemantics();
-    json.arrayEnd();
-    json.removeComma();
+
+    if (global.params.jsonQuery is null)
+    {
+        // Generate the original "non-query format, which is just an array
+        // of modules representing their syntax.
+
+        json.arrayStart();
+        for (size_t i = 0; i < modules.dim; i++)
+        {
+            Module m = (*modules)[i];
+            if (global.params.verbose)
+                message("json gen %s", m.toChars());
+            m.accept(json);
+        }
+        json.arrayEnd();
+        json.removeComma();
+    }
+    else
+    {
+        // Generate the "query format" which is an object where each
+        // output option is it's own field.
+
+        json.objectStart();
+        auto parser = QueryParser(global.params.jsonQuery);
+        for (;;)
+        {
+            parser.skipTrivial();
+            if (*parser.next == '\0')
+                break;
+            auto fieldName = parser.tryParseName();
+            if (fieldName is null)
+            {
+                //json.generateError(format("invalid query: expected name but got '%s' (0x%02x)",
+                //    *parser.next, cast(ubyte)*parser.next));
+                json.generateError("invalid query: expected name or EOF but got something else".ptr);
+                break;
+            }
+            auto fieldID = tryParseJsonField(fieldName);
+            if (!fieldID.hasValue)
+            {
+                json.generateError(("invalid query: unknown field name '" ~ fieldName ~ "'\0").ptr);
+                break;
+            }
+            json.propertyStart((fieldName ~ "\0").ptr);
+            final switch(fieldID.value)
+            {
+                case JsonField.compilerInfo:
+                    json.generateCompilerInfo();
+                    break;
+                case JsonField.buildInfo:
+                    json.generateBuildInfo();
+                    break;
+                case JsonField.modules:
+                    for (size_t i = 0; i < modules.dim; i++)
+                    {
+                        Module m = (*modules)[i];
+                        if (global.params.verbose)
+                            message("json gen %s", m.toChars());
+                            m.accept(json);
+                    }
+                    break;
+                case JsonField.semantics:
+                    json.generateSemantics();
+                    break;
+            }
+        }
+        json.objectEnd();
+    }
+}
+
+/**
+Represents a field that can be generated in the JSON output.
+*/
+enum JsonField
+{
+    compilerInfo,
+    buildInfo,
+    modules,
+    semantics,
+}
+struct OptionalJsonField
+{
+    bool hasValue;
+    JsonField value;
+}
+OptionalJsonField tryParseJsonField(const(char)[] fieldName)
+{
```

Comment:
DRY:
```
foreach(a; __traits(allMembers, JsonField))
  if(fieldName == a) return OptionalJsonField(true, mixin(`JsonField.`~a));
```

Reply:
done, I was thinking I needed to use phobos `EnumMembers`...I forgot I could do this with `__traits`.